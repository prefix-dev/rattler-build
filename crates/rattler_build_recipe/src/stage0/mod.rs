//! The stage0 recipe that contains the un-evaluated recipe information.
//! This means, it still contains Jinja templates and if-else statements.

mod about;
mod build;
pub mod evaluate;
mod extra;
pub mod jinja_functions;
mod package;
mod parser;
mod requirements;
mod source;
mod tests;
mod types;

pub use about::About;
pub use build::Build;
pub use extra::Extra;
pub use package::Package;
pub use parser::{parse_recipe, parse_recipe_from_source};
pub use requirements::Requirements;
pub use source::Source;
pub use tests::TestType;
pub use types::{
    Conditional, ConditionalList, Item, JinjaExpression, JinjaTemplate, ListOrItem, Value,
};

#[derive(Debug, Clone, PartialEq, serde::Serialize)]
pub struct Stage0Recipe {
    pub package: package::Package,
    pub build: build::Build,
    pub requirements: Requirements,
    pub about: about::About,
    pub extra: extra::Extra,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub source: Vec<source::Source>,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub tests: Vec<tests::TestType>,
}

impl Stage0Recipe {
    pub fn used_variables(&self) -> Vec<String> {
        let mut vars = self.package.used_variables();
        vars.extend(self.build.used_variables());
        vars.extend(self.requirements.used_variables());
        vars.extend(self.about.used_variables());
        vars.extend(self.extra.used_variables());
        for src in &self.source {
            vars.extend(src.used_variables());
        }
        for test in &self.tests {
            vars.extend(test.used_variables());
        }
        vars.sort();
        vars.dedup();
        vars
    }
}
