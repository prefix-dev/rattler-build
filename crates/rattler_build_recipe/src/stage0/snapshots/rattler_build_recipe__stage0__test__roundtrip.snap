---
source: crates/rattler_build_recipe/src/stage0/mod.rs
expression: roundtripped
---
package:
  name: conditional-package
  version: 1.5.0
build:
  number: 0
  string: null
  script: {}
  noarch: null
  python:
    entry_points: []
    skip_pyc_compilation: []
    use_python_app_entrypoint: false
    version_independent: false
    site_packages_path: null
  skip: []
  always_copy_files: []
  always_include_files: []
  merge_build_and_host_envs: false
  files: []
  dynamic_linking:
    rpaths: []
    binary_relocation: true
    missing_dso_allowlist: []
    rpath_allowlist: []
    overdepending_behavior: null
    overlinking_behavior: null
  variant:
    use_keys: []
    ignore_keys: []
    down_prioritize_variant: null
  prefix_detection:
    force_file_type:
      text: []
      binary: []
    ignore: false
    ignore_binary_files: false
  post_process: []
requirements:
  build:
  - gcc
  - if: win
    then: vs2019
    else: clang
  - if: unix
    then: make
  host:
  - python
  - if: linux
    then: patchelf
  run:
  - python
  - if: linux
    then: libgcc
  - if: osx
    then: libcxx
  - if: win
    then: msvc-runtime
  - ${{ compiler('c') }}
  - ${{ some_variable }}
  run_exports:
    strong:
    - conditional-package
    - if: cuda
      then: cuda-toolkit
    weak:
    - if: openmp
      then: libgomp
about:
  homepage: https://example.com/
  license: Apache-2.0
  license_family: null
  summary: Package with conditional dependencies
  description: null
  documentation: null
  repository: null
extra:
  recipe-maintainers:
  - maintainer1
